tokio/src/runtime/task/mod.rs

```diff
@@ -50,20 +50,68 @@ unsafe impl<S> Sync for Task<S> {}
 #[repr(transparent)]
 pub(crate) struct Notified<S: 'static>(Task<S>);
 
+/// A task not yet bound to an executor. This object holds two ref-counts to
+/// the task to enable it to be split into two for free.
+pub(crate) struct UnboundTask<T, S: 'static> {
+    raw: RawTask,
+    _p: PhantomData<(T, S)>,
+}
+
+impl<T: 'static + Future, S: 'static + Schedule> UnboundTask<T, S> {
+    /// Bind a scheduler to this task, then split it into two task references.
+    fn bind_and_split(self, scheduler: S) -> (Notified<S>, Task<S>) {
+        // safety: We know what the future type is.
+        let harness = unsafe { Harness::<T, S>::from_raw(self.raw.header().into()) };
+        harness.bind_scheduler(scheduler);
+
+        let task1 = Task {
+            raw: self.raw,
+            _p: PhantomData,
+        };
+
+        let task2 = Task {
+            raw: self.raw,
+            _p: PhantomData,
+        };
+
+        // Don't run destructor.
+        mem::forget(self);
+
+        (Notified(task1), task2)
+    }
+
+    // This method is used by the blocking spawner.
+    pub(crate) fn into_notified(self, scheduler: S) -> Notified<S>
+    where
+        T: Send,
```

Comment:
An unfortunate consequence of this is that we break the fast path in the destructor of join handles for blocking tasks.

Reply:
Could you point to the fast path you are referencing?

Reply:
Yes. The destructor of `JoinHandle` is [here][1]. The implementation of `drop_join_handle_fast` is [here][2]. This CAS will fail because the `into_notified` conversion decrements the ref-count, meaning that the task will no longer be in the initial state when the `JoinHandle` is returned to the user.

[1]: https://github.com/tokio-rs/tokio/blob/8f10d816131868030996b5079fbc632750e69332/tokio/src/runtime/task/join.rs#L241-L251
[2]: https://github.com/tokio-rs/tokio/blob/8f10d816131868030996b5079fbc632750e69332/tokio/src/runtime/task/state.rs#L221-L242